### PR TITLE
DBZ-7272 Guarding against implicit offset invalidation 

### PR DESCRIPTION
**Pull request by jcechace**
https://issues.redhat.com/browse/DBZ-7272

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 1]
```diff
@@ -573,6 +573,14 @@ public static OversizeHandlingMode parse(String value, String defaultValue) {
             .withDescription("Internal use only")
             .withType(Type.LIST);
 
+    /**
+     * The {@link ReplicaSets#SEPARATOR}-separated list of connection strings
+     */
+    public static final Field ALLOW_OFFSET_INVALIDATION = Field.createInternal("mongodb.internal.allow.offset.invalidation")
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:579]:
```suggestion
    public static final Field ALLOW_OFFSET_INVALIDATION = Field.createInternal("mongodb.allow.offset.invalidation")
```
It is already prefixed with `internal`

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java [hunk 2]
```diff
@@ -196,10 +198,49 @@ private MongoDbOffsetContext getPreviousOffset(MongoDbConnectorConfig connectorC
...
+            logger.warn("Found at least one shard specific offset from previous run");
+            if (connectorConfig.isOffsetInvalidationAllowed()) {
+                logger.warn("Offset invalidation is allowed, previous shard specific offsets will be ignored");
+                return;
+            }
+
+            throw new DebeziumException("Found at least one shard specific offset from previous run." +
+                    "The default connection mode for sharded has changed to 'sharded' and previous offsets would be invalidated." +
+                    "Either explicitly set 'mongodb.connection.mode=replica_set' to postpone the migration " +
+                    "or set 'mongodb.internal.allow.offset.invalidation=true'. " +
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java:239]:
```suggestion
                    "or set 'internal.mongodb.allow.offset.invalidation=true'. " +
```

Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java:239]: Please add an information that snapshot will be reexucted.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 1 above)
Comment (jcechace) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:579]: Hm... Then I think the following is not exactly correct either

```
public static final Field TASK_CONNECTION_STRINGS = Field.createInternal("mongodb.internal.task.connection.strings")
            .withDescription("Internal use only")
            .withType(Type.LIST);
```

BTW  wont the change result in the filed named `internal.mongodb.allow.offset.invalidation` then?

Comment (jcechace) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:579]: `TASK_CONNECTION_STRINGS` is however truly internal and is of no concern to the user (as it is internally set). Additionally it's gone in 2.6 so we can safely ignore that :)

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java [hunk 3]
```diff
@@ -196,10 +198,49 @@ private MongoDbOffsetContext getPreviousOffset(MongoDbConnectorConfig connectorC
...
+            logger.warn("Found at least one shard specific offset from previous run");
+            if (connectorConfig.isOffsetInvalidationAllowed()) {
+                logger.warn("Offset invalidation is allowed, previous shard specific offsets will be ignored and snapshot re-executed");
+                return;
+            }
+
+            throw new DebeziumException("Found at least one shard specific offset from previous run." +
+                    "The default connection mode for sharded has changed to 'sharded' and previous offsets would be invalidated." +
+                    "Either explicitly set '" + MongoDbConnectorConfig.CONNECTION_MODE.name() + "=replica_set' to postpone the migration " +
+                    "or set '" + MongoDbConnectorConfig.ALLOW_OFFSET_INVALIDATION.name() + "=true' to re-execute snapshot." +
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java:239]:
```suggestion
                    "or set '" + MongoDbConnectorConfig.ALLOW_OFFSET_INVALIDATION.name() + "=true' to re-execute snapshot and reset offsets." +
```

**Comment by jpechane** [comment]
@jcechace Applied, thanks

**Pull request merged**